graphix/gflow.py

```diff
@@ -461,3 +388,356 @@ def get_layers(l_k):
+
+        Returns
+        -------
+        all_solutions: dict
+            solutions of the linear equations. all_solutions[NODE] is the gflow map of NODE. Note that solutions generally contain uncertainty because the equations are not always square. Therefore, solutions in this step is represented by abstract class. See `SolutionNode` and `find_all_solutions` for details.
+        uncertainty: set
+            set of nodes which are not determined by the linear equations.
+        """
+        # find the rank of the LHS matrix
+        all_solutions = dict()
+
+        for j in range(len(RHS_order)):
+            candidate = RHS_order[j]
+            RHS_vector = RHS[:, j]
+            # perform back substitution
+            all_solutions[candidate] = dict()
+            rank = min(LHS.shape)
+            for row in range(LHS.shape[0] - 1, -1, -1):
+                if row >= rank:
+                    if RHS_vector[row] != 0:
+                        # no solution. empty dict is returned
+                        break
+                    continue
+
+                if LHS[row, row] == 0:
+                    if RHS_vector[row] != 0:
+                        # no solution. empty dict is returned
+                        break
+                    else:
+                        rank -= 1
+                        continue
+                # find the current solution
+                uncertainty = set()
+                for col in range(rank, LHS.shape[1]):
+                    if LHS[row, col] != 0:
+                        uncertainty |= {node_order_col[col]}
+                currect_sol = SolutionNode(node_order_col[row], RHS_vector[row, 0], uncertainty)
+                for col in range(row + 1, rank):
+                    if LHS[row, col] != 0:
+                        currect_sol.subtract(all_solutions[candidate][node_order_col[col]])
+                all_solutions[candidate][node_order_col[row]] = currect_sol
+        uncertainty = {node_order_col[col] for col in range(rank, LHS.shape[1])}
+
+        return all_solutions, uncertainty
+
+
+def find_single_solution(solution, uncertainty):
+    """
+    Find a single gflow solution. This is a special case of `find_all_solutions()`. See `find_all_solutions()` for details.
+
+    Parameters
+    ----------
+    solution: dict
+        solution for specified node.
+    uncertainty: set
+        set of nodes which are not determined by the linear equations.
+
+    Returns
+    -------
+    substituted_solution: set
+        set of nodes which are determined by the uncertainty map.
+    """
+    substituted_solution = set()
+    uncertainty_map = dict(zip(uncertainty, [0] * len(uncertainty)))
+    for solution_node in solution.values():
+        if solution_node.substitute(uncertainty_map):
+            substituted_solution |= {solution_node.node_index}
+    return substituted_solution
+
+
+def find_all_solutions(solutions, uncertainty):
```

Comment:
are you planning to add a method to find N possible solutions rather than all? There might be a case where just a few gflows are useful enough.

Reply:
As much as I understand, here we are adding all possible elements from the  kernel space of the LHS to the solution of our linear equation. Usually a m,n system of rank r has 2^(n-r) solutions.
It might be the case that some of the solutions lead to patterns that are close to each other in terms of the unitaries they might implement (once we designate a measurement plane ), but I think it is still beneficial to have an indea of all possible gflows for a graph atleast for the sake of generality.

Did I understand it correctly @masa10-f ?

Reply:
@pafloxy thanks for the comment! I agree it's useful to find all indeed. 
I was just wondering whether it's useful to add optional method to _not_ find all but a few, because Masato mentioned it takes quite long to find all solutions.

Reply:
I see. I think I would agree that to make it *optional*, given all solutions here refers to exponentially many in the size of the nullspace of the matrix it would slow down the algorithm considerably.

Reply:
Sorry for the late reply. I was busy last week.
Thank you for your discussions. I'm afraid that we don't have good criteria for which N gflow to select. So, I implemented another option called 'abstract' that returns abstract nodes and uncertainty nodes of each layer. Though users will need to manually substitute uncertainty map as in `find_single_solution` or `find_all_solutions`, this approach offers flexibility in finding suitable solutions.